// @progress/kendo-react-treeview/dist/npm/utils/getItemIdUponKeyboardNavigation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var kendo_react_common_1 = require("@progress/kendo-react-common");
var kendo_react_common_2 = require("@progress/kendo-react-common");
/**
 * @hidden
 */
function getItemIdUponKeyboardNavigation(item, itemId, items, keyCode, fieldsSvc) {
    switch (keyCode) {
        case kendo_react_common_1.Keys.left:
            return getIdUponLeftKey();
        case kendo_react_common_1.Keys.right:
            return getIdUponRightKey();
        case kendo_react_common_1.Keys.up:
            return getIdUponUpKey();
        case kendo_react_common_1.Keys.down:
            return getIdUponDownKey();
        case kendo_react_common_1.Keys.home:
            return kendo_react_common_2.treeIdUtils.ZERO_LEVEL_ZERO_NODE_ID;
        case kendo_react_common_1.Keys.end:
            return getIdUponEndKey();
        default:
            return itemId;
    }
    function getIdUponLeftKey() {
        return kendo_react_common_2.treeIdUtils.isIdZeroLevel(itemId)
            // For expanded disabled items, navigate directly to the parent.
            // Collapse cannot happen because an event is not fired to the client.
            || (fieldsSvc.expanded(item) && (0, kendo_react_common_2.isEnabledAndAllParentsEnabled)(itemId, items, fieldsSvc)) ?
            itemId : kendo_react_common_2.treeIdUtils.getDirectParentId(itemId);
    }
    function getIdUponRightKey() {
        return (0, kendo_react_common_2.isItemExpandedAndWithChildren)(item, fieldsSvc) ? kendo_react_common_2.treeIdUtils.getFirstChildId(itemId) : itemId;
    }
    function getIdUponUpKey() {
        var index = Number(kendo_react_common_2.treeIdUtils.getShortId(itemId));
        var parentId = kendo_react_common_2.treeIdUtils.getDirectParentId(itemId);
        return index ? getLastNotOpenedChildId(kendo_react_common_2.treeIdUtils.createId(index - 1, parentId), items, fieldsSvc) :
            kendo_react_common_2.treeIdUtils.isIdZeroLevel(itemId) ? itemId : parentId;
    }
    function getIdUponDownKey() {
        if ((0, kendo_react_common_2.isItemExpandedAndWithChildren)(item, fieldsSvc)) {
            return kendo_react_common_2.treeIdUtils.getFirstChildId(itemId);
        }
        else {
            return getNextItemId(itemId, items, fieldsSvc) || itemId;
        }
    }
    function getIdUponEndKey() {
        var result = (items.length - 1).toString();
        var currentItem = items[items.length - 1];
        var children;
        while ((0, kendo_react_common_2.isItemExpandedAndWithChildren)(currentItem, fieldsSvc)) {
            children = currentItem[fieldsSvc.getChildrenField()];
            result = kendo_react_common_2.treeIdUtils.createId(children.length - 1, result);
            currentItem = children[children.length - 1];
        }
        return result;
    }
}
exports.default = getItemIdUponKeyboardNavigation;
function getNextItemId(itemId, items, fieldsSvc) {
    var parentId = kendo_react_common_2.treeIdUtils.getDirectParentId(itemId);
    var parent = parentId ? kendo_react_common_2.treeIdUtils.getItemById(parentId, items, fieldsSvc.getChildrenField()) : undefined;
    var siblings = parent ? parent[fieldsSvc.getChildrenField()] : items;
    var index = Number(kendo_react_common_2.treeIdUtils.getShortId(itemId));
    return index < siblings.length - 1 ?
        kendo_react_common_2.treeIdUtils.createId(index + 1, parentId) :
        parent ? getNextItemId(parentId, items, fieldsSvc) : undefined;
}
function getLastNotOpenedChildId(itemId, items, fieldsSvc) {
    var item = kendo_react_common_2.treeIdUtils.getItemById(itemId, items, fieldsSvc.getChildrenField());
    if ((0, kendo_react_common_2.isItemExpandedAndWithChildren)(item, fieldsSvc)) {
        return getLastNotOpenedChildId(kendo_react_common_2.treeIdUtils.createId(item[fieldsSvc.getChildrenField()].length - 1, itemId), items, fieldsSvc);
    }
    else {
        return itemId;
    }
}
