// @progress/kendo-react-common/dist/npm/tree-utils/FieldsService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TreeFieldsService = void 0;
var misc_1 = require("./misc");
/**
 * @hidden
 */
var TreeFieldsService = /** @class */ (function () {
    function TreeFieldsService(treeViewProps) {
        this.expandField = treeViewProps.expandField;
        this.selectField = treeViewProps.selectField;
        this.hasChildrenField = treeViewProps.hasChildrenField;
        this.childrenField = treeViewProps.childrenField;
        this.textField = treeViewProps.textField;
        this.disableField = treeViewProps.disableField;
        this.checkField = treeViewProps.checkField;
        this.checkIndeterminateField = treeViewProps.checkIndeterminateField;
        this.focusIdField = treeViewProps.focusIdField;
    }
    TreeFieldsService.prototype.expanded = function (item) {
        return (0, misc_1.getNestedValue)(this.expandField, item);
    };
    TreeFieldsService.prototype.selected = function (item) {
        return (0, misc_1.getNestedValue)(this.selectField, item);
    };
    TreeFieldsService.prototype.text = function (item) {
        return (0, misc_1.getNestedValue)(this.textField, item);
    };
    TreeFieldsService.prototype.disabled = function (item) {
        return (0, misc_1.getNestedValue)(this.disableField, item);
    };
    TreeFieldsService.prototype.hasChildren = function (item) {
        return (0, misc_1.getNestedValue)(this.hasChildrenField, item);
    };
    TreeFieldsService.prototype.children = function (item) {
        var children = (0, misc_1.getNestedValue)(this.childrenField, item);
        return children ? children : [];
    };
    TreeFieldsService.prototype.checked = function (item) {
        return (0, misc_1.getNestedValue)(this.checkField, item);
    };
    TreeFieldsService.prototype.checkIndeterminate = function (item) {
        return (0, misc_1.getNestedValue)(this.checkIndeterminateField, item);
    };
    TreeFieldsService.prototype.focusId = function (item) {
        return this.focusIdField && (0, misc_1.getNestedValue)(this.focusIdField, item);
    };
    TreeFieldsService.prototype.getChildrenField = function () {
        return this.childrenField;
    };
    return TreeFieldsService;
}());
exports.TreeFieldsService = TreeFieldsService;
