// @progress/kendo-react-treeview/dist/es/utils/getItemIdUponKeyboardNavigation.js
import { Keys } from '@progress/kendo-react-common';
import { isItemExpandedAndWithChildren, isEnabledAndAllParentsEnabled, treeIdUtils } from '@progress/kendo-react-common';
/**
 * @hidden
 */
export default function getItemIdUponKeyboardNavigation(item, itemId, items, keyCode, fieldsSvc) {
    switch (keyCode) {
        case Keys.left:
            return getIdUponLeftKey();
        case Keys.right:
            return getIdUponRightKey();
        case Keys.up:
            return getIdUponUpKey();
        case Keys.down:
            return getIdUponDownKey();
        case Keys.home:
            return treeIdUtils.ZERO_LEVEL_ZERO_NODE_ID;
        case Keys.end:
            return getIdUponEndKey();
        default:
            return itemId;
    }
    function getIdUponLeftKey() {
        return treeIdUtils.isIdZeroLevel(itemId)
            // For expanded disabled items, navigate directly to the parent.
            // Collapse cannot happen because an event is not fired to the client.
            || (fieldsSvc.expanded(item) && isEnabledAndAllParentsEnabled(itemId, items, fieldsSvc)) ?
            itemId : treeIdUtils.getDirectParentId(itemId);
    }
    function getIdUponRightKey() {
        return isItemExpandedAndWithChildren(item, fieldsSvc) ? treeIdUtils.getFirstChildId(itemId) : itemId;
    }
    function getIdUponUpKey() {
        var index = Number(treeIdUtils.getShortId(itemId));
        var parentId = treeIdUtils.getDirectParentId(itemId);
        return index ? getLastNotOpenedChildId(treeIdUtils.createId(index - 1, parentId), items, fieldsSvc) :
            treeIdUtils.isIdZeroLevel(itemId) ? itemId : parentId;
    }
    function getIdUponDownKey() {
        if (isItemExpandedAndWithChildren(item, fieldsSvc)) {
            return treeIdUtils.getFirstChildId(itemId);
        }
        else {
            return getNextItemId(itemId, items, fieldsSvc) || itemId;
        }
    }
    function getIdUponEndKey() {
        var result = (items.length - 1).toString();
        var currentItem = items[items.length - 1];
        var children;
        while (isItemExpandedAndWithChildren(currentItem, fieldsSvc)) {
            children = currentItem[fieldsSvc.getChildrenField()];
            result = treeIdUtils.createId(children.length - 1, result);
            currentItem = children[children.length - 1];
        }
        return result;
    }
}
function getNextItemId(itemId, items, fieldsSvc) {
    var parentId = treeIdUtils.getDirectParentId(itemId);
    var parent = parentId ? treeIdUtils.getItemById(parentId, items, fieldsSvc.getChildrenField()) : undefined;
    var siblings = parent ? parent[fieldsSvc.getChildrenField()] : items;
    var index = Number(treeIdUtils.getShortId(itemId));
    return index < siblings.length - 1 ?
        treeIdUtils.createId(index + 1, parentId) :
        parent ? getNextItemId(parentId, items, fieldsSvc) : undefined;
}
function getLastNotOpenedChildId(itemId, items, fieldsSvc) {
    var item = treeIdUtils.getItemById(itemId, items, fieldsSvc.getChildrenField());
    if (isItemExpandedAndWithChildren(item, fieldsSvc)) {
        return getLastNotOpenedChildId(treeIdUtils.createId(item[fieldsSvc.getChildrenField()].length - 1, itemId), items, fieldsSvc);
    }
    else {
        return itemId;
    }
}
